Add overloads for from_pretrained and from_dict on PretrainedConfig

ringohoffman opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Related #23980

this is part of a set of changes to improve the type checking in `PreTrainedModel.from_pretrained`

1. Pull `return_unused_kwargs` out of `kwargs` to create `overload`s for `PretrainedConfig.from_pretrained` & `PretrainedConfig.from_dict` for improved return typing
2. add type hints to class variables in `FlaxPreTrainedModel`, `TFPreTrainedModel`, and `PreTrainedModel` so that we can consume the above improvements

before:
<img width="656" alt="Screenshot 2023-06-03 at 10 48 18 PM" src="https://github.com/huggingface/transformers/assets/27844407/d435f97a-bd13-40e4-90f7-43287927aedb">

after:
<img width="660" alt="Screenshot 2023-06-03 at 10 47 31 PM" src="https://github.com/huggingface/transformers/assets/27844407/c0a9e781-6d42-4a44-a3be-ce05c1589b0c">


## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Documentation: @sgugger, @stevhliu and @MKhalusova

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

Commit:
Add overloads for from_pretrained and from_dict on PretrainedConfig

also type class variables for improved type hinting in FlaxPreTrainedModel, TFPreTrainedModel, and PreTrainedModel

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_23983). All of your documentation changes will be reflected on that endpoint.

Commit:
Change name_or_path property to use getattr again but change default to be a str

my original intent in changing it was to respect the indicated return type since it doesn't seem like anywhere that accesses name_or_path is ever expecting it to be None...

sgugger reviewed (commented):
Same comments as for the two other PRs. Those changes hurt the code readability.
